libevmasm/Assembly.cpp

```diff
@@ -791,10 +792,20 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
 				iter = eliminator.feedItems(iter, m_items.end(), usesMSize);
 				bool shouldReplace = false;
 				AssemblyItems optimisedChunk;
+				auto runGas = [&](AssemblyItems const& items) {
+					GasMeter gasMeter{std::make_shared<KnownState>(eliminator.getKnownState()), _settings.evmVersion};
+					GasMeter::GasConsumption gas;
+					for (auto const& item: items)
+						gas += gasMeter.estimateMax(item);
+					return gas;
+				};
 				try
 				{
 					optimisedChunk = eliminator.getOptimizedItems();
-					shouldReplace = (optimisedChunk.size() < static_cast<size_t>(iter - orig));
+					shouldReplace = (
+						optimisedChunk.size()  < static_cast<size_t>(iter - orig) &&
+						!(runGas(AssemblyItems(orig, iter)) < runGas(optimisedChunk))
```

Comment:
@ekpyron , Although this makes sense, this change is not related to `PUSH0`. At first I thought the problem we were having was due to this, but I checked and only 4 tests are affected by this (gas cost reduced by 1 or 2):

 - `semanticTests/abiEncoderV2/calldata_overlapped_dynamic_arrays.sol`
 - `semanticTests/array/copying/array_copy_target_leftover.sol`
 - `semanticTests/events/event_dynamic_array_storage.sol`
 - `semanticTests/events/event_dynamic_array_storage_v2.sol`

I was thinking we could merge this in a separate PR for organization sake, but in any case, it's just a small change.

Reply:
The proper way to decide would also take the "runs" parameter into account like e.g. the low-level inliner does (see https://github.com/ethereum/solidity/blob/c0ef06c362f86e1e35c4663e27913e1cabd3f5e6/libevmasm/Inliner.cpp#L195 and the calculation leading to it)

Reply:
But then wouldn't the `runs` parameter be on both sides of the inequality? Because, both the block being replaced and the replacement would be executed as often during the contract lifetime. 
In that inliner case, the cost of executing the function, inlined or not, is omitted based on the same argument (and also something I thought when working on this).
The parameter "runs" is used, because in the not-inlined case, besides de function body, there is the additional cost of the `CallSitePattern` and `functionPattern`.

Reply:
I mean, this is generally a rather inexact science, since applying an optimization may lead to further optimizations, which may then increase the overall metric, even though the first step seemingly made things worse - so the main question is what changes in the measured results.
The condition you have now, improves the costs in the tests compared to the previous code-size-only metric, right?

Reply:
I think I understand it better now.


It improves, but marginally. The tests I've listed have insignificant gas reduc

> While the condition you have now (both need to improve) fails, right?

Right, both have to be improved. I see now the point that the "runs" param is needed.





> The condition you have now, improves the costs in the tests compared to the previous code-size-only metric, right?

Yeah, but only on those 4 listed tests above, and insignificant amounts of gas. Maybe they can be even related to the problem you pointed later. I will verify the results after making "runs" be considered in the gas cost comparison.